testsrc/org/mozilla/javascript/tests/MessageProviderLanguageTest.java

```diff
@@ -0,0 +1,113 @@
+    private final Locale NOT_TRANSLATED = new Locale("gd");
+
+    private final String messageId = "msg.invalid.date";
+    private final String englishMessage = "Date is invalid.";
+    private final String frenchMessage = "Date incorrecte.";
+    private final String defaultMessage = englishMessage;
+
+    private Context cx;
+
+    @Before
+    public void init() {
+        cx = Context.enter();
+    }
+
+    @After
+    public void cleanup() {
+        Context.exit();
+    }
+
+    @Test
+    public void defaultLocaleUsesDefaultLanguageAndContextLocaleIsNotSpecified() {
+        assertEquals(englishMessage, getMessageUsingDefaultLocale(US));
+    }
+
+    @Test
+    public void defaultLocaleUsesOtherTranslatedLanguageAndContextLocaleIsNotSpecified() {
+        assertEquals(frenchMessage, getMessageUsingDefaultLocale(FRANCE));
+    }
+
+    @Test
+    public void defaultLocaleUsesUntranslatedLanguageAndContextLocaleIsNotSpecified() {
+        assertEquals(defaultMessage, getMessageUsingDefaultLocale(NOT_TRANSLATED));
+    }
+
+    @Test
+    public void defaultLocaleUsesDefaultLanguageAndContextLocaleIsTheSame() {
+        assertEquals(englishMessage, getMessageUsingContextLocale(US, US));
+    }
+
+    @Test
+    public void defaultLocaleUsesDefaultLanguageAndContextLocaleUsesOtherTranslatedLanguage() {
+        assertEquals(frenchMessage, getMessageUsingContextLocale(FRANCE, US));
+    }
+
+    @Test
+    public void defaultLocaleUsesDefaultLanguageAndContextLocaleIsNotTranslated() {
+        assertEquals(englishMessage, getMessageUsingContextLocale(NOT_TRANSLATED, US));
+    }
+
+    @Test
+    public void defaultLocaleUsesOtherTranslatedLanguageAndContextLocaleIsTheSame() {
+        assertEquals(frenchMessage, getMessageUsingContextLocale(FRANCE, FRANCE));
+    }
+
+    @Test
+    public void defaultLocaleUsesOtherTranslatedLanguageAndContextLocaleUsesDefaultLanguage() {
+        assertEquals(englishMessage, getMessageUsingContextLocale(US, FRANCE));
+    }
+
+    @Test
+    public void defaultLocaleUsesOtherTranslatedLanguageAndContextLocaleIsNotTranslated() {
+        assertEquals(frenchMessage, getMessageUsingContextLocale(NOT_TRANSLATED, FRANCE));
+    }
+
+    @Test
+    public void defaultLocaleIsNotTranslatedAndContextLocaleIsTheSame() {
+        assertEquals(defaultMessage, getMessageUsingContextLocale(NOT_TRANSLATED, NOT_TRANSLATED));
+    }
+
+    @Test
+    public void defaultLocaleIsNotTranslatedAndContextLocaleUsesDefaultLanguage() {
+        assertEquals(englishMessage, getMessageUsingContextLocale(US, NOT_TRANSLATED));
+    }
+
+    @Test
+    public void defaultLocaleIsNotTranslatedAndContextLocaleUsesOtherTranslatedLanguage() {
+        assertEquals(frenchMessage, getMessageUsingContextLocale(FRANCE, NOT_TRANSLATED));
+    }
+
+    private String getMessageUsingDefaultLocale(Locale defaultLocale) {
+        Locale.setDefault(defaultLocale);
```

Comment:
I think a test should not change the system status. What about saving the current locale and restore in a finally block.

Reply:
Yes, thanks! We already have a number of tests that don't clean up the environment properly so we have to fork them into separate JVMs, which means that the test run takes much longer.